### DISCUSSION TITLE
Custom cache invalidation logic

### DISCUSSION THREAD

**Discussion by mohsen1** [apollographql/apollo-server#4305] [General]
We are using `@cacheControl` and it works great! However a time based cache expiration is not enough for us. We want to be able to invalidate specific cache keys when a mutation is successful. 

Consider following API:

```graphql
type Person {
  id: ID
}

type Team {
  members: [Person!]!
}

extend type Query {
  team: Team  @cacheControl(maxAge: 10000)
}

extend type Mutation {
  addMemeber(): Team
}
```

When I invoke `addMember` I would like to invalidate the cache for `Query:team`. How can I implement such logic?

I thought about a custom directive that specifies which cache keys should be invalidated but I'm not quite sure how to design that:

```graphql
extend type Mutation {
  addMemeber(): Team @invalidateCache(paths: ["Query:team"])
}
```

**Comment by ralexrdz** [comment]
looking for the same feature or a workaround

**Comment by intellix** [comment]
This product just appeared from the guys behind the playground: https://graphcdn.io/ which invalidates the cache after a mutation and stores/invalidates in CDN edges. 

I came here to ask that a similar feature get baked directly into Apollo Cache as it's currently the reason we only set maxAge to 5 seconds, to avoid stale data after mutations. We want cache, but we can't deal with the fact that it's not possible to invalidate that at all.

Something like Apollo Client InMemory cache alone would be awesome, which is an in-memory state which gets updated after mutations and subscriptions using the typename + id

**Reply by geosigno** [reply]
+1

**Comment by matthewattanasio** [comment]
Any update on the baked-in feature to invalidate a CDN cache after a mutation, I am using CloudFront...

**Reply by geosigno** [reply]
Go with [graphcdn.io](https://graphcdn.io/), this is a game changer ;)

**Reply by matthewattanasio** [reply]
They do look like they have it all figured out ;)
